File: linked_list.py
# Full implementation of the linked list algorithm in the Python programming language

# Node class
class Node:
    # Constructor
    def __init__(self, data):
        self.data = data
        self.next = None

# Linked list class


class LinkedList:
    # Constructor
    def __init__(self):
        self.head = None

    # Insert a node at the beginning of the linked list
    def insert_first(self, data):
        # Create a new node to be inserted first
        new_node = Node(data)
        # Point the new node to the head
        new_node.next = self.head
        # Change the pointer of the new node to the head to adjust links
        self.head = new_node

    # Insert after a certain node in the linked list
    def insert_after(self, prev_node, data):
        # Check if the previous node is equal to None
        if not prev_node:
            print('The previous node cannot be null')
            return
        # Create a new node to be inserted after the previous one
        new_node = Node(data)
        # Point the new node to the next node after the previous one
        new_node.next = prev_node.next
        # Point the previous node to the new node to change links
        prev_node.next = new_node

    # Insert the last node in the linked list
    def insert_last(self, data):
        # Create a new node to be inserted last
        new_node = Node(data)
        # Check if there's no head in the linked list, point the new node to the head and return
        if not self.head:
            self.head = new_node
            return
        # Traverse the linked list until the last node points to None, point the last node to the new node
        temp = self.head
        while (temp.next):
            temp = temp.next
        temp.next = new_node

    # Delete a node
    def delete_node(self, position):
        # Check if there's no head
        if not self.head:
            return
        # Check if the deleted node is in the head on the linked list, move the head pointer to the next node, thus deleting it from the liked list
        temp = self.head
        if position == 0:
            self.head = temp.next
            temp = None
        # Find the key to be deleted from the linked list
        for i in range(position - 1):
            temp = temp.next
            if temp == None:
                return

        # Check if the key is not present in the linked list
        if not temp:
            return
        if not temp.next:
            return
        # Delete a node
        next = temp.next.next
        temp.next = None
        temp.next = next

  # Search for a node in the linked list
    def search_node(self, key):
        # Set the current node to the head to get started with
        current = self.head
        # Traverse the linked list till the end and compare data on each iteration to the key that's passed in as an argument to the method
        while (current):
            if current.data == key:
                return True
            current = current.next
        return False

    # Sort linked list
    def sort_list(self):
        # Set the current node to the head to get started with
        current = self.head
        # Set the index that points to the next node to null at the beginning
        index = None
        # Check if there's no head in the linked list
        if not self.head:
            return
        # Traverse the linked list and on each iteration compare the value of the current node to the value of the next node
        # If the current value is greater than the next one, swap the two adjacent nodes
        while (current):
            index = current.next
            while (index):
                if current.data > index.data:
                    current.data, index.data = index.data, current.data
                index = index.next
            current = current.next

    # Print the linked list
    def print_list(self):
        # Traverse the linked list till the end and print values on each iteration
        temp = self.head
        while (temp):
            print(f'Linked list: {temp.data} ', end='')
            temp = temp.next
        print()


if __name__ == '__main__':
    # Instantiate the linked list class => create an object
    ll = LinkedList()
    ll.insert_first(10)
    ll.insert_first(20)
    ll.insert_first(50)
    ll.insert_after(ll.head.next, 120)
    ll.insert_after(ll.head.next.next, 160)
    ll.insert_last(209)
    ll.insert_first(306)
    ll.delete_node(5)
    searched_node = 305
    if ll.search_node(searched_node):
        print(f'The node with value {searched_node} has been found')
        print()
    else:
        print(f'The node with the value {searched_node} has not been found')
        print()
    ll.print_list()
    ll.sort_list()
    ll.print_list()

File: linked_list/linked_list.cpp
/* Here is a full implementation of singly linked list algorithm in the C++ language */
#include <iostream>

// Define Node structure
struct Node
{
    int data;
    Node *next;
};

/* Function forward declaration & prototyping */

// Insert at the beginning of the linked list
void insertFirst(Node **headdRef, int data);

// Insert after a particular node
void insertAfter(Node *prevNode, int data);

// Insert last
void insertLast(Node **headRef, int data);

// Delete node
void deleteNode(Node **headRef, int key);

// Search for a node
bool searchNode(Node **headRef, int key);

// Sort list
void sortList(Node **headRef);

// Print the linked list
void printList(Node *head);

int main()
{
    // At the beginning the list is empty and the head points to null
    Node *head = NULL;
    insertFirst(&head, 50);
    insertFirst(&head, 100);
    insertFirst(&head, 200);
    insertLast(&head, 750);
    insertAfter(head->next, 350);
    deleteNode(&head, 350);
    deleteNode(&head, 750);
    deleteNode(&head, 960);
    insertFirst(&head, 567);
    insertLast(&head, 990);
    insertAfter(head->next->next, 3968);
    insertFirst(&head, 5690);
    sortList(&head);
    int key = 200;
    if (searchNode(&head, key))
    {
        std::cout << "The node " << key << " has been found" << std::endl;
    }
    else
    {
        std::cout << "The node " << key << " has not been found" << std::endl;
    }
    printList(head);
    return 0;
}

/* Function definition */

void insertFirst(Node **headRef, int data)
{
    // Dynamically allocate memory on the heap for a new node
    Node *newNode = new Node;
    // Assign values
    newNode->data = data;
    // Point the new node to the head
    newNode->next = *headRef;
    // Shift the head pointer to the new node
    *headRef = newNode;
}

void insertAfter(Node *prevNode, int data)
{
    // Make sure the previous node exists
    if (prevNode == NULL)
    {
        std::cout << "The previous node can't be null" << std::endl;
        return;
    }
    // Dynamically allocate memory on the heap for a new node
    Node *newNode = new Node;
    // Assign values
    newNode->data = data;
    // Point the new node to the node next to the previous one
    newNode->next = prevNode->next;
    // Point the previous node to the new node to change links
    prevNode->next = newNode;
}

void insertLast(Node **headRef, int data)
{
    // Create temp variable to traverse the linked list
    Node *temp = *headRef;
    // Dynamically allocate memory on the heap for a new node
    Node *newNode = new Node;

    // Check if the head is empty, then assign the node to the head and return
    if (*headRef == NULL)
    {
        *headRef = newNode;
        return;
    }
    // Assign values
    newNode->data = data;
    newNode->next = NULL;

    // Traverse the linked list till the end and point the last node to the new node
    while (temp->next != NULL)
    {
        temp = temp->next;
    }
    temp->next = newNode;
}

void deleteNode(Node **headRef, int key)
{
    // Check if the node to be deleted is in the head, point the head to the next position
    Node *temp = *headRef, *prev;
    if (temp->next != NULL && temp->data == key)
    {
        *headRef = temp->next;
        delete temp;
    }
    // Traverse the linked list and find the key to delete
    while (temp != NULL && temp->data != key)
    {
        prev = temp;
        temp = temp->next;
    }
    // Check if there's no match in the list
    if (temp == NULL)
    {
        std::cout << "There's no node to be deleted" << std::endl;
        return;
    }
    // Delete the node
    prev->next = temp->next;
}

bool searchNode(Node **headRef, int key)
{
    // Check if there's no head
    if (headRef == NULL)
    {
        return false;
    }
    // Make the current variable to traverse the linked list
    Node *current = *headRef;
    while (current)
    {
        if (current->data == key)
        {
            return true;
        }
        current = current->next;
    }
    return false;
}

void printList(Node *head)
{
    // Temp variable to traverse the list
    Node *temp = head;
    while (temp != NULL)
    {
        std::cout << "Linked list " << temp->data << " ";
        temp = temp->next;
    }
    std::cout << std::endl;
}

void sortList(Node **headRef)
{
    Node *current = *headRef;
    Node *index = NULL;
    int temp;

    // Check if there's no head
    if (headRef == NULL)
    {
        return;
    }
    // Implement a swapping algorithm
    while (current)
    {
        index = current->next;
        while (index)
        {
            if (current->data > index->data)
            {
                temp = current->data;
                current->data = index->data;
                index->data = temp;
            }
            index = index->next;
        }
        current = current->next;
    }
}
